sdks/python/apache_beam/transforms/external.py

```diff
@@ -42,6 +46,97 @@
   grpc = None
 # pylint: enable=wrong-import-order, wrong-import-position, ungrouped-imports
 
+DEFAULT_EXPANSION_SERVICE = 'localhost:8097'
+
+
+def iter_urns(coder, context=None):
+  yield coder.to_runner_api_parameter(context)[0]
+  for child in coder._get_component_coders():
+    for urn in iter_urns(child, context):
+      yield urn
+
+
+class External(ptransform.PTransform):
+  """
+  Base class for transforms that produce external transforms.
+  """
+  _urn = None
+  _schema = None
+
+  def __init__(self, expansion_service=None):
+    self.expansion_service = expansion_service or DEFAULT_EXPANSION_SERVICE
+
+  @classmethod
+  def config_value(cls, obj, typehint=None):
+    """
+    Helper to create a ConfigValue with an encoded value.
+    """
+    if typehint is None:
+      coder = registry.get_coder(type(obj))
+    else:
+      coder = registry.get_coder(typehint)
+    return ConfigValue(
+      coder_urn=list(iter_urns(coder)),
+      payload=coder.encode(obj))
+
+  @classmethod
+  def encode_config(cls, config):
+    schema = cls._schema or {}
+    result = {}
+    for k, v in config.items():
+      typehint = schema.get(k)
+      if v is None and (
+          typehint is None or not isinstance(typehint, typehints.Optional)):
+        # make it easy for user to filter None by default
+        continue
+      result[k] = cls.config_value(v, typehint)
+    return result
+
+  def get_config(self):
+    """
+    Get a dictionary of configuration values to encode.
+
+    This provides a default implementation that determines the config
+    from the class's schema dictionary. If using this implementation,
+    the schema's keys are expected to correspond to instance
+    attributes on the External subclass.
+
+    :return: dictionary of configuration values
+    """
+    if self._schema is None:
+      raise RuntimeError("External subclasses must implement a schema "
+                         "dictionary to use automatic encoding")
+    config = {}
+    missing = []
+    for key in self._schema:
+      try:
+        config[key] = getattr(self, key)
+      except AttributeError:
+        missing.append(key)
+    return config
+
+  def get_encoded_config(self):
```

Comment:
Do you expect user's to override (a subset of) all these methods ? Can we reduce the public surface of this class by just keeping get_encoded_config() (for maximum flexibility) and get_schema() (instead of cls._schema) public and making other public methods (get_config, encode_config, config_value) start with _ ?

Reply:
>  Do you expect user's to override (a subset of) all these methods ?

The idea is to make the `_schema` design as clean and easy to use as possible, so that most users only need to implement a schema.  But I still wanted to provide the flexibility to override methods to modify the behavior for advanced users, if desired.  

> Can we reduce the public surface of this class by just keeping get_encoded_config() (for maximum flexibility) and get_schema() (instead of cls._schema) public and making other public methods (get_config, encode_config, config_value) start with _ ?

Sure, that works for me.  

My ultimate goal is the `dataclasses` design that I outlined above, and it should be very easy to write a `get_schema()` implementation that derives the schema dict from a `dataclass`-based `ExternalTransform`.

Reply:
Addressed in my latest push.

Reply:
To clarify my `dataclasses` comment: that's for a future PR, and it won't require `dataclasses`.

Reply:
Thanks. Looks updated public API good.